Fixing JIT compatibility

azrael417 opened the pull request:
Fixing JIT compatibility by changing process group vs None comparison to proper type without implicit type casting.

Fixes not yet reported bug,

azrael417 committed:
Fixing JIT compatibility

Fixing JIT compatibility by changing process group vs None comparison to proper type without implicit type casting.

albanD commented:
I'm not sure that is ok to do?
Since `self.process_group` is `Any`. This is changing the behavior of that if statement as things like False or empty string will not be considered False anymore.

azrael417 commented:
The comparison 

if self.process_group

is not OK for torch jit. What about demoting it to Optional[ProcessGroup] (or whatever the type is) instead of Any?

jbschlosser commented:
> I'm not sure that is ok to do? Since `self.process_group` is `Any`. This is changing the behavior of that if statement as things like False or empty string will not be considered False anymore.

`process_group` is immediately passed to `get_world_size()` after the if statement, and I think non-`ProcessGroup` objects like `False` / empty string would error out there. Not sure why the typing is `Optional[Any]` now, but it seems like it can be changed to `Optional[ProcessGroup]` without disrupting much in practice.

azrael417 committed:
Update batchnorm.py

Fixed typing in SyncBatchNorm

azrael417 commented:
Hi all,

thanks for the helpful comments. I have updated the file accordingly, please let me know what you think.

Best 
Thorsten

jbschlosser commented on torch/nn/modules/batchnorm.py:
```diff
@@ -10,6 +10,7 @@
 from .lazy import LazyModuleMixin
 from .module import Module
 
+from torch._C._distributed_c10d import ProcessGroup
```
Hm maybe this is why the typing was `Optional[Any]` before. Won't this be a problem if PyTorch is built with `USE_DISTRIBUTED=0` ?

jbschlosser replied:
Follow-up question: why is `SyncBatchNorm` not in distributed :p

albanD replied:
Don't know...

azrael417 replied:
Shall we refactor that into distributed? Or what is the best way to go about it?

albanD replied:
I think refactoring it would be quite a bit of work.
I think that for this PR you can move back to `optional[Any]`. Sorry for the back and forth.

azrael417 replied:
But still do the is not None comparison?

azrael417 replied:
Yes, that is right.

azrael417 replied:
OK, I undid the changes. Next thing on the list is how to deal with the CLA. My company has signed it so I cannot sign it individually, what do I need to do?

azrael417 committed:
Update batchnorm.py

Rolling back previous changes, i.e. making process_group argument to syncBN Any type again.

ptrblck commented:
Is this test failure real?
```
RuntimeError: 
Unknown custom class type c10d.ProcessGroup. Please ensure it is registered.:
c10d::broadcast(__torch__.torch.classes.c10d.ProcessGroup _0, Tensor[] _1, int _2, int _3, int _4) -> __torch__.torch.classes.c10d.Work _0
                                             ~~~~~~~~~~~~ <--- HERE
```

albanD commented:
This would need to be rebased on top of latest master to be sure
